### DISCUSSION TITLE
[Feature Request][input-date] add localized datemasks

### DISCUSSION THREAD

**Discussion by DawnCarrasco** [ing-bank/lion#90] [General]
`lion-date` has a validation message that says `Please enter a valid date (DD/MM/YYYY).` but is checking for `MM/DD/YYYY`. The localisation that we are using is `en`. Thanks!

![image](https://user-images.githubusercontent.com/137844/59257864-15e61180-8c37-11e9-9b16-0ef9f1b0e728.png)

## Short-term solution (done)

Use a specific dialect, e.g. `en-PH` in this case, with a proper date format in translations.

## Long-term solution (to be implemented)

We should stop hardcoding date format in validation messages and dynamically set it based on the locale value. We can provide a special variable in the context of validators translations.

@tlouisse has a POC:

https://github.com/ing-bank/lion/tree/poc/localized-datemasks
- https://github.com/ing-bank/lion/blob/poc/localized-datemasks/packages/validate/src/validators.js#L3
- https://github.com/ing-bank/lion/blob/poc/localized-datemasks/packages/localize/src/date/getDateMask.js

**Comment by bashmish** [comment]
I believe that bug is in the error message. It should clearly say that the expected format is `MM/DD/YYYY`, while we have a wrong error message in `@lion/validate/translations/en.js`.

The reason you have an US format is because you are using a shorthand locale `en` which assumes US standard for dates, while the recommended approach is to use a full locale always like `lang-COUNTRY`, and to the best of my knowledge you need `en-GB`, so this will be a quick fix for you.

We are going to eventually deprecate shorthand locales btw, exactly because of issues like this.

**Comment by donecarlo** [comment]
> I believe that bug is in the error message. It should clearly say that the expected format is `MM/DD/YYYY`, while we have a wrong error message in `@lion/validate/translations/en.js`.
> 
> The reason you have an US format is because you are using a shorthand locale `en` which assumes US standard for dates, while the recommended approach is to use a full locale always like `lang-COUNTRY`, and to the best of my knowledge you need `en-GB`, so this will be a quick fix for you.
> 
> We are going to eventually deprecate shorthand locales btw, exactly because of issues like this.

Can we use `en-PH`? Or we need to create a PR for validation messages first?

**Comment by bashmish** [comment]
@carlocanlas yes, you need `en-PH.js` file there which either extends from `en.js` ([like here](https://github.com/ing-bank/lion/blob/%40lion/validate%400.2.7/packages/validate/translations/en-AU.js)) or defines the needed translations from scratch (like in `en.js` itself). Basically you need the file to export all the keys with proper translations, so you need to look into what you extends from and make sure it is all correct in the basic object and only extends if smth is different in the Philippine English.

Some insights about the localisation system design are [documented here](https://github.com/ing-bank/lion/tree/master/packages/localize#usage-for-component-developers).

**Comment by bashmish** [comment]
While we now support en-PH, this issue is a different thing and also needs a fix.

This happens due to the fact that the platform's Intl uses the format `MM/DD/YYYY` for the locale `en`, while we have for it the error message with `DD/MM/YYYY`.

We have a few options to fix it:
1. Forbid the usages of generic locales like `en` in favor of dialect specific ones like `en-GB`. (@LarsDenBakker  suggested that)
2. Fix translations in `en.js` so that they include `MM/DD/YYYY`, and fix all other files so that they still have a proper fallback with `DD/MM/YYYY`.
3. Remove hardcoded formats from translations and make a special variable available in translations which contains the format applicable to a specific locale. (@tlouisse [has a POC](https://github.com/ing-bank/lion/tree/poc/localized-datemasks) for this which is not as simple as it seems because M, D and Y need to be translated to each language and preferable loaded asynchronously)

I personally like the solution 3, because it will remove the burden of hardcoding all these values in our translations.

At the same time this will be fixed as a side effect of solution 1 if we implement it soon.

**Comment by tlouisse** [comment]
For completeness, this was the poc @bashmish was referring to: https://github.com/ing-bank/lion/tree/poc/localized-datemasks
See:  
- https://github.com/ing-bank/lion/blob/poc/localized-datemasks/packages/validate/src/validators.js#L3
- https://github.com/ing-bank/lion/blob/poc/localized-datemasks/packages/localize/src/date/getDateMask.js

Keep in mind it is just a poc and it should be tackled in an improvement story of the Validation System later

**Comment by bashmish** [comment]
I updated the description and the title of this issue to reflect the agreement of our team regarding the proper fix for this.

**Comment by jorenbroekema** [comment]
Hi, since this will be fixed when we fix https://github.com/ing-bank/lion/issues/187 , I am closing this one as a duplicate of that issue.

**Comment by jorenbroekema** [comment]
Reopening as a feature enhancements for localized datemasks

**Comment by erikkroes** [comment]
Hi,

We are closing this issue because we're changing the way we handle feature requests. Your issue will get the "votes needed" label, and people can add 👍 reactions to [vote and show interest](https://github.com/ing-bank/lion/issues?q=label%3A%22votes+needed%22+sort%3Areactions-%2B1-desc). It will be reopened when this feature gets picked up. 

This we way we clearly focus on bugs in our open issues. At the same time we can see what community interest there is for new features.

Thanks!
